pilot/pkg/serviceregistry/kube/controller/ambient/ambientindex.go

```diff
@@ -138,21 +138,43 @@ func New(options Options) Index {
 		ObjectTransform: kubeclient.StripPodUnusedFields,
 	}, krt.WithName("Pods"))
 
+	// TODO: Should this go ahead and transform the full ns into some intermediary with just the details we care about?
+	Namespaces := krt.NewInformer[*v1.Namespace](options.Client, krt.WithName("Namespaces"))
+
 	MeshConfig := MeshConfigCollection(ConfigMaps, options)
 	Waypoints := WaypointsCollection(Gateways)
-	WaypointIndex := krt.CreateIndex[Waypoint, model.WaypointScope](Waypoints, func(w Waypoint) []model.WaypointScope {
-		// We can be a part of a service account waypoint, or a namespace waypoint
-		return []model.WaypointScope{{Namespace: w.Namespace, ServiceAccount: w.ForServiceAccount}}
-	})
 
 	// AllPolicies includes peer-authentication converted policies
 	AuthorizationPolicies, AllPolicies := PolicyCollections(AuthzPolicies, PeerAuths, MeshConfig)
 	AllPolicies.RegisterBatch(PushXds(a.XDSUpdater, func(i model.WorkloadAuthorization) model.ConfigKey {
 		return model.ConfigKey{Kind: kind.AuthorizationPolicy, Name: i.Authorization.Name, Namespace: i.Authorization.Namespace}
 	}), false)
 
-	WorkloadServices := a.ServicesCollection(Services, ServiceEntries)
+	// these are workloadapi-style services combined from kube services and service entries
+	WorkloadServices := a.ServicesCollection(Services, ServiceEntries, Waypoints, Namespaces)
 	ServiceAddressIndex := krt.CreateIndex[model.ServiceInfo, networkAddress](WorkloadServices, networkAddressFromService)
+	ServiceInfosByOwningWaypoint := krt.CreateIndex[model.ServiceInfo, networkAddress](WorkloadServices, func(s model.ServiceInfo) []networkAddress {
+		// Filter out waypoint services
+		if s.Labels[constants.ManagedGatewayLabel] == constants.ManagedGatewayMeshControllerLabel {
+			return nil
+		}
+		waypoint := s.Waypoint
+		if waypoint == nil {
+			return nil
+		}
+		waypointAddress := waypoint.GetAddress()
+		if waypointAddress == nil {
+			return nil
+		}
+
+		ip := waypointAddress.GetAddress()
```

Comment:
Would we ever have a case where a waypoint might have more than one address? e.g. dualstack?

Does it make sense to key `ServiceInfosByOwningWaypoint` with the entire `Waypoint` struct, rather than extracting the zeroth address from inside the `Waypoint` struct and using that?

Reply:
Agreed. We've had some strong requirements about what is/isn't valid for waypoint addressing. We will need to adjust those as we bring in capabilities for sandwiched waypoints from non-istio gateways. IMO that would be a better time/place to adjust the logic instead of expanding the scope of this PR.

Reply:
The other case would be if we decide not to use a Service to track the waypoint instances and instead materialize them directly on the Status field. Could be important in clusters where there was ClusterIP starvation.